pilot_experiment/functions/process_data.py

```diff
@@ -0,0 +1,179 @@
+        input_data_file,
+        header=0,
+        sep='\t',
+        index_col=0).T
+
+    # Read metadata
+    metadata = pd.read_table(
+        metadata_file,
+        header=0,
+        sep='\t',
+        index_col=0)
+        
+    map_experiment_sample = metadata[['sample_name', 'ml_data_source']]
+    
+    # Get expression data associated with experiment_id
+    selected_mapping = map_experiment_sample.loc[lst_experiments]
+    selected_sample_ids = list(selected_mapping['ml_data_source'].values)
+
+    selected_data = normalized_data.loc[selected_sample_ids]
+
+    print(selected_data.shape)
+    print(selected_data.head(10))
+
+    # Save selected gene expression data
+    selected_data.to_csv(output_file, sep='\t', index=True)
+
+def permute_expression_data(input_data_file,
+output_file):
+    """
+    Returns permuted version of input expression data to be used as a baseline
+    for downstream analysis
+
+    Arguments
+    ----------
+    input_data_file: str
+        File containing normalized gene expression data
+
+        ------------------------------| PA0001 | PA0002 |...
+        05_PA14000-4-2_5-10-07_S2.CEL | 0.8533 | 0.7252 |...
+        54375-4-05.CEL                | 0.7789 | 0.7678 |...
+        ...                           | ...    | ...    |...
+
+    output_file: str
+        Filename to save permuted expression data
+    """
+    # Read data
+    selected_data = pd.read_table(
+        input_data_file,
+        header=0,
+        sep='\t',
+        index_col=0)
+
+    # For each sample, shuffle gene values (i.e. shuffle values independently per sample) in order to
+    # disrupt any gene structure
+    shuffled_arr = []
+    num_samples = selected_data.shape[0]
+
+    for i in range(num_samples):
+        row = list(selected_data.values[i])
+        shuffled_row = random.sample(row, len(row))
+        shuffled_arr.append(shuffled_row)
+
+    shuffled_selected_data = pd.DataFrame(shuffled_arr,
+                                        index=selected_data.index,
+                                        columns=selected_data.columns)
+
+    print(shuffled_selected_data.head(10))
+
+    # Save permuted gene expression data
+    shuffled_selected_data.to_csv(output_file, sep='\t', index=True)
+
+def annotate_genes(input_data_file,
+gene_annotation_file,
+output_file):
+    """
+    Returns file mapping gene ids to label ('core' or 'accessory') using gene annotations from
+    BACTOME annotation tool: https://pseudomonas-annotator.shinyapps.io/pa_annotator/
+
+     Arguments
+    ----------
+    input_data_file: str
+        File containing normalized gene expression data
+
+        ------------------------------| PA0001 | PA0002 |...
+        05_PA14000-4-2_5-10-07_S2.CEL | 0.8533 | 0.7252 |...
+        54375-4-05.CEL                | 0.7789 | 0.7678 |...
+        ...                           | ...    | ...    |...
+
+    gene_annotation_file: str
+        File containing mapping between PAO1 and PA14 gene ids
```

Comment:
This was something that I had wrong for a while so don't worry. the name does indeed use the letter "O" but the gene ids use "0" (i.e. PA0001). Its definitely confusing because you would expect it to use the number "0" especially since there is a PA14, but thats not the case!